src/main/java/bio/terra/service/snapshot/flight/delete/DeleteSnapshotPopAndLockDatasetStep.java

```diff
@@ -0,0 +1,94 @@
+import bio.terra.service.snapshot.exception.SnapshotNotFoundException;
+import bio.terra.service.snapshot.flight.SnapshotWorkingMapKeys;
+import bio.terra.stairway.FlightContext;
+import bio.terra.stairway.FlightMap;
+import bio.terra.stairway.Step;
+import bio.terra.stairway.StepResult;
+import bio.terra.stairway.StepStatus;
+import java.util.UUID;
+
+public class DeleteSnapshotPopAndLockDatasetStep implements Step {
+
+  private final SnapshotService snapshotService;
+  private final DatasetService datasetService;
+  private final UUID snapshotId;
+  private final AuthenticatedUserRequest authenticatedUserRequest;
+  private final boolean sharedLock;
+
+  public DeleteSnapshotPopAndLockDatasetStep(
+      UUID snapshotId,
+      SnapshotService snapshotService,
+      DatasetService datasetService,
+      AuthenticatedUserRequest authenticatedUserRequest,
+      boolean sharedLock) {
+    this.snapshotId = snapshotId;
+    this.snapshotService = snapshotService;
+    this.datasetService = datasetService;
+    this.authenticatedUserRequest = authenticatedUserRequest;
+    this.sharedLock = sharedLock;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext context) {
+    FlightMap map = context.getWorkingMap();
+    Snapshot snapshot;
+
+    // Confirm that snapshot exists and populate info about snapshot to inform optional steps
+    try {
+      snapshot = snapshotService.retrieve(snapshotId);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_EXISTS, true);
+      boolean hasGoogleProject =
+          snapshot.getProjectResource() != null
+              && snapshot.getProjectResource().getGoogleProjectId() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_GOOGLE_PROJECT, hasGoogleProject);
+      boolean hasAzureStorageAccount = snapshot.getStorageAccountResource() != null;
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_AZURE_STORAGE_ACCOUNT, hasAzureStorageAccount);
+    } catch (SnapshotNotFoundException snapshotNotFoundException) {
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_EXISTS, false);
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, false);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_GOOGLE_PROJECT, false);
+      map.put(SnapshotWorkingMapKeys.SNAPSHOT_HAS_AZURE_STORAGE_ACCOUNT, false);
+      return StepResult.getStepResultSuccess();
+    }
+
+    // Now we've confirmed the snapshot exists, let's check on the source dataset
+    UUID datasetId = snapshot.getFirstSnapshotSource().getDataset().getId();
+    map.put(DatasetWorkingMapKeys.DATASET_ID, datasetId);
+    try {
+      datasetService.lock(datasetId, context.getFlightId(), sharedLock);
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, true);
+    } catch (DatasetNotFoundException notFoundEx) {
+      map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, false);
+    } catch (RetryQueryException | DatasetLockException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
```

Comment:
This is fine, although it might be more obvious that the `put()` is always called if the code had a boolean that it used to store the state, so the `put()` can be done outside the `try`.
```suggestion
    boolean datasetExists = true;
    try {
      datasetService.lock(datasetId, context.getFlightId(), sharedLock);
    } catch (DatasetNotFoundException notFoundEx) {
      datasetExists = false;
    } catch (RetryQueryException | DatasetLockException e) {
      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
    }
    map.put(SnapshotWorkingMapKeys.DATASET_EXISTS, datasetExists);
```